src/html.rs

```diff
@@ -51,34 +54,81 @@ pub fn simplify(html: &str, lang: &str) -> String {
             }
         }
 
-        for id in to_remove.drain(..) {
-            if let Some(mut node) = document.tree.get_mut(id) {
-                node.detach();
-            }
-        }
+        remove_ids(&mut document, to_remove.drain(..));
     } else {
         warn!("No sections to remove configured for lang {lang:?}");
     }
 
-    // Remove elements with no text that isn't whitespace.
+    // Handle remaining removals in single pass.
 
-    for element in document
+    for el in document
         .root_element()
         .descendants()
         .filter_map(ElementRef::wrap)
     {
-        if element.text().all(|t| t.trim().is_empty()) {
-            to_remove.push(element.id());
+        if is_image(&el) || is_empty_or_whitespace(&el) {
+            to_remove.push(el.id());
         }
     }
+    remove_ids(&mut document, to_remove.drain(..));
+
+    // Final transformations.
 
-    for id in to_remove.drain(..) {
+    let base_url = Url::parse(&format!("https://{}.wikipedia.org/wiki/", &lang)).unwrap();
+    fix_relative_urls(&mut document, base_url);
+
+    document.html()
+}
+
+fn remove_ids(document: &mut Html, ids: impl IntoIterator<Item = NodeId>) {
+    for id in ids {
         if let Some(mut node) = document.tree.get_mut(id) {
             node.detach();
         }
     }
+}
 
-    document.html()
+fn is_empty_or_whitespace(el: &ElementRef) -> bool {
+    el.text().all(|t| t.trim().is_empty())
```

Comment:
Is it possible to check whitespaces without modification? E.g. use is_whitespace ?

Reply:
We could do `el.text().flat_map(str::chars).all(char::is_whitespace)`. I was going to say that working with characters might be less efficient, since the `Pattern`s can [work directly in UTF-8](https://doc.rust-lang.org/src/core/str/pattern.rs.html#374), but it looks like [the implementation of `trim`](https://doc.rust-lang.org/src/core/str/mod.rs.html#1867) also uses `char::is_whitespace`.

Reply:
As trim actually returns a slice, your implementation is already optimal )

Reply:
One thing I realized though is that `trim` still has to check the right side if it encounters non-whitespace characters from the left. `trim_left` is better, but both still need to construct the slice. I don't think the compiler can optimize that away.

I tried the three approaches with a selection of strings and found that the char iterator was fastest. Ultimately a micro-optimization but still interesting :wink:.
 
```rust
const MIXED: &str =          " \tcd\nfg ";
const NOT_WHITESPACE: &str = "abcdefgh";
const WHITESPACE: &str =     " \t\n \t\t  ";
```

```
running 9 tests
test chars_mixed              ... bench:           3 ns/iter (+/- 0)
test chars_not_whitespace     ... bench:           0 ns/iter (+/- 0)
test chars_whitespace         ... bench:           7 ns/iter (+/- 1)
test trim_left_mixed          ... bench:           4 ns/iter (+/- 1)
test trim_left_not_whitespace ... bench:           3 ns/iter (+/- 0)
test trim_left_whitespace     ... bench:           9 ns/iter (+/- 0)
test trim_mixed               ... bench:          10 ns/iter (+/- 1)
test trim_not_whitespace      ... bench:           5 ns/iter (+/- 4)
test trim_whitespace          ... bench:          12 ns/iter (+/- 0)
```

Reply:
This is an important outcome of an interesting project: to learn something new )